neqo-http3/src/push_controller.rs

```diff
@@ -25,15 +96,333 @@ impl Display for PushController {
 }
 
 impl PushController {
-    #[allow(clippy::needless_pass_by_value)]
-    pub fn new_push_promise(&self, push_id: u64, header_block: Vec<u8>) -> Res<()> {
+    /// A new `push_promise` has been received.
+    /// # Errors
+    /// `HttpId` if `push_id` greater than it is allowed has been received.
+    pub fn new_push_promise(
+        &mut self,
+        push_id: u64,
+        ref_stream_id: u64,
+        header_block: Vec<u8>,
+    ) -> Res<()> {
         qtrace!(
             [self],
-            "New push promise push_id={} header_block={:?}",
+            "New push promise push_id={} header_block={:?} max_push={}",
             push_id,
-            header_block
+            header_block,
+            self.max_concurent_push
         );
         qtrace!("A new push promise {} {:?}", push_id, header_block);
-        Err(Error::HttpId)
+
+        self.check_push_id_and_create_stream_state(push_id)?;
+
+        match self.push_streams.get_mut(&push_id) {
+            None => {
+                qtrace!(
+                    "Push has been closed already {} {}.",
+                    self.next_push_id_to_open,
+                    push_id
+                );
+                Ok(())
+            }
+            Some(push_state) => match push_state {
+                PushState::Init => {
+                    self.conn_events
+                        .push_promise(push_id, ref_stream_id, header_block.clone());
+                    *push_state = PushState::PushPromise {
+                        headers: header_block,
+                    };
+                    Ok(())
+                }
+                PushState::PushPromise { headers } | PushState::Active { headers, .. } => {
+                    if header_block != *headers {
```

Comment:
Are you asserting that the headers are exactly identical?  That's going to take a fair bit of work, right?

I say that because you will be creating an entry on line 117, and then comparing it against itself, which isn't ideal.  Can that drop out early?

Reply:
Line 117 creates only an entry in Init state and in that state we do not compare headers. It is never compared against itself.

These headers are not decompress yet (Issue #342 ). This will not work correctly until the issue is fixed.